cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/AppMetadataStore.java

```diff
@@ -478,40 +480,84 @@ public Map<ApplicationId, ApplicationMeta> getApplicationsForAppIds(Collection<A
-      }
-      try (JsonReader reader = new JsonReader(new StringReader(appMeta))) {
-        reader.beginObject();
-        while (reader.peek() != JsonToken.END_OBJECT) {
-          String name = reader.nextName();
-          if (name.equals("spec")) {
-            existingPrograms.addAll(ApplicationSpecificationAdapter.getProgramIds(appId, reader));
-          } else {
-            reader.skipValue();
+    // The latest version programs
+    Map<ApplicationReference, Set<ProgramId>> latestProgramIdsMap = new HashMap<>();
+    // The latest versions
+    Map<ApplicationReference, String> latestAppVersions = new HashMap<>();
+    try (CloseableIterator<StructuredRow> iterator =
+           getApplicationSpecificationTable().multiScan(multiRanges, appRefs.size() * 2)) {
+      while (iterator.hasNext()) {
+        StructuredRow row = iterator.next();
+        ApplicationId appId = getApplicationIdFromRow(row);
+        String appMeta = row.getString(StoreDefinition.AppMetadataStore.APPLICATION_DATA_FIELD);
+        if (appMeta == null) {
+          throw new IOException("Missing application metadata for application " + appId);
+        }
+        
+        ApplicationReference appRef = appId.getAppReference();
+        String isLatest = row.getString(StoreDefinition.AppMetadataStore.LATEST_FIELD);
+        // Get either the latest versioned or "-SNAPSHOT" programs.
+        if (!Objects.equals(isLatest, "true") && latestProgramIdsMap.containsKey(appRef)) {
+          // Only proceed when it's latest, or we didn't add this AppReference
+          // Otherwise skip current application
+          continue;
+        }
+        Set<ProgramId> currentProgramsSet = new HashSet<>();
+        try (JsonReader reader = new JsonReader(new StringReader(appMeta))) {
+          reader.beginObject();
+          while (reader.peek() != JsonToken.END_OBJECT) {
+            String name = reader.nextName();
+            if (name.equals("spec")) {
+              currentProgramsSet.addAll(ApplicationSpecificationAdapter.getProgramIds(appId, reader));
+            } else {
+              reader.skipValue();
+            }
           }
+          reader.endObject();
         }
-        reader.endObject();
+        latestProgramIdsMap.put(appRef, currentProgramsSet);
+        latestAppVersions.put(appRef, appId.getVersion());
       }
     }
-    return programIds.stream().filter(existingPrograms::contains).collect(Collectors.toSet());
+
+    List<ProgramId> versionedProgramIds = new ArrayList<>();
+
+    // Put in the latest version into programId
+    for (ProgramReference programRef : programRefs) {
+      ApplicationReference appRef = programRef.getParent();
+      if (latestAppVersions.containsKey(appRef)) {
+        ProgramId actualProgramId = appRef
+          .app(latestAppVersions.get(appRef))
+          .program(programRef.getType(), programRef.getProgram());
+        versionedProgramIds.add(actualProgramId);
+      } else {
```

Comment:
if it's not in latestAppVersions, doesn't that mean it doesn't exist? Why would we return it in that situation rather than filtering it out? (previous behavior looks like it would filter it out)